client/extensions/woocommerce/app/products/product-variation-types-form.js

```diff
@@ -0,0 +1,198 @@
+/**
+ * External dependencies
+ */
+import React, { Component, PropTypes } from 'react';
+import i18n from 'i18n-calypso';
+
+/**
+ * Internal dependencies
+ */
+import FormLabel from 'components/forms/form-label';
+import FormTextInput from 'components/forms/form-text-input';
+import Button from 'components/button';
+import TokenField from 'components/token-field';
+
+export default class ProductVariationTypesForm extends Component {
+
+	static propTypes = {
+		product: PropTypes.object.isRequired,
+		editProduct: PropTypes.func.isRequired,
+		addVariation: PropTypes.func.isRequired,
+	};
+
+	constructor( props ) {
+		super( props );
+
+		this.addVariationType = this.addVariationType.bind( this );
+		this.updateVariations = this.updateVariations.bind( this );
+		this.updateType = this.updateType.bind( this );
+		this.updateValues = this.updateValues.bind( this );
+	}
+
+	getNewFields() {
+		return {
+			type: '',
+			values: [],
+		};
+	}
+
+	addVariationType( event ) {
+		event.preventDefault();
+		const updatedVariations = [ ...this.props.product.variationTypes, this.getNewFields() ];
+		this.props.editProduct( null, 'variationTypes', updatedVariations );
+	}
+
+	updateType( index, event ) {
+		event.preventDefault();
+		const updatedVariations = [ ...this.props.product.variationTypes ];
+		updatedVariations[ index ] = { ...updatedVariations[ index ], type: event.target.value };
+		this.props.editProduct( null, 'variationTypes', updatedVariations );
+	}
+
+	updateValues( index, value ) {
+		const updatedVariations = [ ...this.props.product.variationTypes ];
+		updatedVariations[ index ] = { ...updatedVariations[ index ], values: value };
+		this.props.editProduct( null, 'variationTypes', updatedVariations );
+	}
+
+	/*
+	 * Given a set of variationTypes and a product's exisiting variations,
+	 * return new variation entires for those that need to be created.
+	 * Both variationTypes and currentVariations should come from the product object.
+	 * The format of the variations with their attribute data matches the output of
+	 * the REST API.
+	 */
+	generateVariations( variationTypes = [], currentVariations = [] ) {
+		const possibleValues = [];
+		const indexesToLabels = [];
+		const possibleVariations = [];
+		const existingVariations = [];
+		const variations = [];
+
+		/*
+		 * Create an array of each set of possible values for a specific type
+		 * so that we can generate a new cartesian of products.
+		 * We also need to store a relation of indexes to the type label so we can
+		 * properly build new variation objects later.
+		*/
+		variationTypes.map( function( variationType ) {
```

Comment:
could use `forEach` instead of `map` since you're only iterating for side-effects, not actually using the return.